`WidgetState` mapping

nate-thegrate opened the pull request:
This pull request implements [enhanced enum](https://dart.dev/language/enums#declaring-enhanced-enums) features for the new `WidgetState` enum, in order to improve the developer experience when creating and using `WidgetStateProperty` objects.

`WidgetState` now has a `.matchesSet()` method:

```dart
// identical to "states.contains(WidgetState.error)"
final bool hasError = WidgetState.error.isSatisfiedBy(states);
```

This addition allows for wide variety of `WidgetStateProperty` objects to be constructed in a simple manner.

<br><br>

```dart
// before
final style = MaterialStateTextStyle.resolveWith((states) {
  if (states.contains(MaterialState.error)) {
    return const TextStyle(color: Colors.red);
  } else if (states.contains(MaterialState.focused)) {
    return const TextStyle(color: Colors.blue);
  }
  return const TextStyle(color: Colors.black);
});


// after
const style = WidgetStateTextStyle.fromMap({
  WidgetState.error:   TextStyle(color: Colors.red),
  WidgetState.focused: TextStyle(color: Colors.blue),
  WidgetState.any:     TextStyle(color: Colors.black),
});
```

```dart
// before
final color = MaterialStateProperty.resolveWith((states) {
  if (states.contains(MaterialState.focused)) {
    return Colors.blue;
  } else if (!states.contains(MaterialState.disabled)) {
    return Colors.black;
  }
  return null;
});


// after
final color = WidgetStateProperty<Color?>.fromMap({
  WidgetState.focused:   Colors.blue,
  ~WidgetState.disabled: Colors.black,
});
```

```dart
// before
const activeStates = [MaterialState.selected, MaterialState.focused, MaterialState.scrolledUnder];

final color = MaterialStateColor.resolveWith((states) {
  if (activeStates.any(states.contains)) {
    if (states.contains(MaterialState.hovered) {
      return Colors.blueAccent;
    }
    return Colors.blue;
  }
  return Colors.black;
});


// after
final active = WidgetState.selected | WidgetState.focused | WidgetState.scrolledUnder;

final color = WidgetStateColor.fromMap({
  active & WidgetState.hovered: Colors.blueAccent,
  active:  Colors.blue,
  ~active: Colors.black,
});
```

<br>

(fixes #146042, and also fixes #143488)

cedvdb commented:
imo `WidgetState.isError(states)` would be more intuitive than `WidgetState.error(states)`

nate-thegrate commented:
@cedvdb this is a really good point.

In order to make the mapping work, I [added a `call()` method](https://github.com/nate-thegrate/flutter/blob/widget-state-mapping/packages/flutter/lib/src/widgets/widget_state.dart#L156) to `WidgetState`, and it has the nice side effect of providing a syntax that's more concise than `states.contains()`. Since it's a `call()` method for the enum value rather than a static class method, this would be a change to the enum name:

```dart
enum WidgetState { isHovered, isFocused, isPressed, ... isError }
```

<br>

Luckily, there's already a way to add an intuitive `states.isError` syntax, as mentioned in a [Flutter YouTube video](https://youtu.be/CylXr3AF3uU?t=449). And with this PR, there's no need to directly reference `states` at all:

```dart
WidgetStateProperty.map({WidgetState.error: Colors.red, WidgetState.any: Colors.black});
```

MitchellGoodwin reviewed (commented):
Thank you for this PR! I think this is a great solution, that covers simple cases and complicated combinations well. I have some questions I'd like to talk about, but no complaints otherwise.

MitchellGoodwin reviewed (approved):
LGTM with one follow up on naming. Thanks again for this PR, it's a great addition.

nate-thegrate commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -171,8 +244,8 @@ abstract class WidgetStateColor extends Color implements WidgetStateProperty<Col
 
-class _WidgetStateColor extends WidgetStateColor {
-  _WidgetStateColor(this._resolve) : super(_resolve(_defaultStates).value);
+class _WidgetStateColorWith extends WidgetStateColor {
```
I'd never thought about how the `with` keyword could make this confusing. Sounds great!

Hixie commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -9,6 +9,64 @@ import 'package:flutter/services.dart';
+  bool call(Set<WidgetState> states);
+}
+
+/// A private class, used in [WidgetStateOperators].
```
trivial nit: this should be `//` not `///` per the style guide, because we wouldn't want to publish the documentation in its current state if we made the class public. see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-for-public-quality-private-documentation

Hixie commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -9,6 +9,64 @@ import 'package:flutter/services.dart';
+  }
+}
+
+/// A private class, used to create [WidgetState.any].
```
same here

MitchellGoodwin commented:
The Google testing looks related. In one spot, there is a set of only a few WidgetStates.
Pseudo code it looks like this.
```
const certainStates = <WidgetState>{
   WidgetState.pressed,
   Widget.State.hovered,
}
```

It threw this error over this:
Error: Constant evaluation error: const certainStates = <WidgetState>{

Context: The element 'WidgetState {}' conflicts with another existing element in the set.
 - 'WidgetState' is from 'package:flutter/src/widgets/widget_state.dart'
 WidgetState.focused,
 
 Context: While analyzing: const interactiveStates = <WidgetState>{

MitchellGoodwin commented:
It's possible there are other issues, but at least most of them seem to be around that set

nate-thegrate commented:
Dang. Guess I'll stop trying to repeatedly run the checks 🙂

MitchellGoodwin commented:
This PR when merged (#148804) is going to add a similar `const` set of `WidgetState`s that should throw the same error as what's in Google testing. We would need to create a way to allow having a constant set of states, because that would be a bit of a bummer of a regression.

nate-thegrate commented:
> We would need to create a way to allow having a constant set of states, because that would be a bit of a bummer of a regression.

@MitchellGoodwin I have some good news: `const Set<WidgetState>` still works!

Just to double-check, I checked out my branch and did a `flutter run` with the following:

```dart
// main.dart

void main() {
  runApp(const MaterialApp(home: Scaffold()));

  const WidgetStateMap<String> widgetStateMap = {
    WidgetState.selected: 'it works!',
  };
  const Set<WidgetState> states = {
    WidgetState.selected,
    WidgetState.hovered,
  };

  final doesItWork = WidgetStateProperty.map(widgetStateMap).resolve(states);
  print(doesItWork); // it works!
}
```

I'm looking forward to #148804 being merged; at that point hopefully we can copy those changes into this branch and see what's going on.

nate-thegrate committed:
implement `WidgetState` mapping

nate-thegrate commented:
Super random, but I wanted to double-check that `WidgetStateProperty.map` was a `const` constructor (since `const` WidgetStateMaps will probably be a common occurrence), and it turns out that it wasn't!

<br>

As far as Google Testing errors, we could continue waiting for #148804, but this has been on my mind for a while and I think I might have an idea.

If a `const` set of states were causing the issues, then the button tests would also fail:

https://github.com/flutter/flutter/blob/c63339866d4f7477777e7377bf044f14b743f71c/packages/flutter/test/material/filled_button_test.dart#L2256-L2258

My best guess is that it's an import error: now that the enum has been changed on my branch, maybe the test file is trying to import 2 different versions of the same type at once. (I ran into this error before when I made a class that had a name overlap with the Material library, and then tried to import both `material.dart` and the file I made.)

A super quick & dirty fix would be to give each import a different prefix to guarantee that there won't be any overlaps.

(Also, I feel like I should apologize for this attempt to boss you around. I really wish I could just see the test output directly and try to fix it myself, but I guess it is what it is.)

MitchellGoodwin commented:
> My best guess is that it's an import error: now that the enum has been changed on my branch, maybe the test file is trying to import 2 different versions of the same type at once.

Took a look through. It's not a unit test that's failing but a build error on a smoketest. The set that's being a problem is called `interactiveStates`, which we do use in some tests and examples, but they are scopped in a way so I do not think they would conflict.

Looking at the stack trace, that set is first called for this kind of function:
```final numberOfCurrentInteractions =
states.intersection(interactiveStates).length;
```
One file imports Material, the other one, the Widgets library. Neither import much more, both files are high level.

nate-thegrate commented:
Dang, I guess I'll try digging around on the internet for answers.

I really appreciate you looking into it.

nate-thegrate commented:
I think I was able to replicate the error message:

```
lib/test.dart:3:27: Error: Constant evaluation error:
const interactiveStates = {
                          ^
lib/test.dart:5:15: Context: The element 'WidgetState {}' conflicts with another existing element in the set.
 - 'WidgetState' is from 'package:flutter/src/widgets/widget_state.dart' ('/C:/…/widget_state.dart').
  WidgetState.focused,
              ^
lib/test.dart:3:7: Context: While analyzing:
const interactiveStates = {
      ^
```

```dart
// test.dart

import 'package:flutter/material.dart';

const interactiveStates = {
  WidgetState.focused,
  WidgetState.focused,
};
```

I have no clue how it relates to this PR.

MitchellGoodwin commented:
@nate-thegrate Google testing is failing on different tests then before, but they look odd. Can you rebase so we can see if it's not related to something else? One of the tests is a golden file test with a difference under 0.1% so I can't see what's different, and the other looks more like a difference in backend responses that I wouldn't think is related to WidgetState. If it fails again I'll dig deeper.

MitchellGoodwin commented:
Looks like it's the same errors as before, with the constant evaluation error.

goderbauer commented:
This is a very clever implementation! However, I am concerned that it makes `WidgetState` and the code it is used in harder to comprehend. If you see code like `states.contains(MaterialState.error)` it's immediately clear what's happening. That's not true for `WidgetState.error(states)`. It's not obvious at all that the latter returns a boolean, and if I want to figure out what this does it gets even more confusing: At first glance, I would have expected that `WidgetState.error` is some kind of a static function, but if I click through I confusingly end up at an enum value - and I am sort of stuck there. There is no indication what "invoking" an enum value would get me. I have to remember the dart magic trick that implementing a `call` method makes any object callable and then look at the implementation of said `call` method to figure out what's actually going on in this code snippet. In the original (`states.contains(MaterialState.error)`) it's clear right there what's happening. To understand the new world (`WidgetState.error(states)`) I have to do two hops (one of which requires specialized knowledge about dart's magic - and rarely used - `call` method) to get an understanding of what the code does.

This is also true for the proposed `map` method: It's very compact and clever! But it's also way harder to understand what's happening there.

tl;dr: I admire the technical skills and clever implementation behind this. It's genius! Ultimately, I think we shouldn't do this, though, as it will hurt readability and understandability of the code.

nate-thegrate commented:
@goderbauer I really appreciate this feedback!

I received similar feedback for #148449, and I was happy to close it (since I was feeling uneasy about it as well).
Likewise, after reading your thoughts about this PR, I agree that the implementation is severely lacking in readability and understandability.

However: rather than close this right away, I'd like to take some time this evening to work through the issues you mentioned.
I fervently believe that an improved version of this PR would be immensely useful for the people who want a more straightforward `WidgetPropertyResolver` implementation.

nate-thegrate commented:
When I look at enum values in VS Code, they have consistent syntax highlighting, which creates a clear visual distinction between a "call" method and a static class method.

Unfortunately, other places (such as GitHub) don't behave the same way. And even if they did, I admit it was still was a bad idea to use `call()` here, due to its impact on readability.

<br>

I've updated the enum to use a named method `matchesSet`, to prevent the appearance of directly invoking the object.
I think it makes the new constructor's `resolve` method significantly easier to follow:

```dart
for (final MapEntry<WidgetStateMapKey, T> entry in map.entries) {
  if (entry.key.matchesSet(states)) {
    return entry.value;
  }
}
```

<br>

I also gave the [`matchesSet` doc comments](https://github.com/flutter/flutter/blob/0afd932a69e10fce59150a393ecdbe856a3eb20f/packages/flutter/lib/src/widgets/widget_state.dart#L22-L39) a clearer explanation:

> Whether this key should apply, given the current [states].
> 
> If the key is a single [WidgetState] object, it matches the set if the set contains the object.
> 
> The key could also be created using one or more operators, for example:
> ```dart
> final WidgetStateMapKey mapKey = WidgetState.focused | WidgetState.hovered;
> ```
> 
> In the above case, the output of `mapKey.matchesSet(states)` is equivalent to:
> ```dart
> states.contains(WidgetState.focused) || states.contains(WidgetState.hovered);
> ```

<br>

I believe that the name change and added documentation enhances the readability and maintainability of the `WidgetStateMap` interface.

nate-thegrate commented:
### ✨ 🙏 Let us sing Hosanna, for miracles cometh from on high 🙏 ✨

<br>

I do not understand it whatsoever, but here's what I've discovered:
- Adding an operator to `WidgetState` → works (but doesn't allow chaining operators)
- Adding operators to both `WidgetStateMapKey` and `WidgetState` → doesn't work
- Putting operators in a mixin, applied to both classes → doesn't work
- Making an `extension on WidgetStateMapKey` → works!

This would have been much more straightforward if enums could extend other classes, but I guess we gotta work with what we got.

I didn't want to throw on `autosubmit` right away: extensions are used in multiple places throughout the repo, but the [style guide](https://github.com/flutter/flutter/blob/fa70e610cc244750b07321c466385d821533ee68/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-extension) recommends against them.

It isn't possible to subclass an enum, so the only way that this extension would make someone upset is if they implemented a different subclass of `WidgetStateMapKey` and they didn't like the operator logic.
In light of the fact that Google Testing wonkiness makes an `extension` the only working option here, I believe this is a situation where it's warranted. (If someone could give a thumbs-up or something, that would be much appreciated.)

goderbauer commented:
Awesome. Before we submit this, I'd like to take another look at this from a readability and comprehension point of view. It's in my queue and I hope to get to it either this or early next week.

nate-thegrate commented:
> Before we submit this, I'd like to take another look at this from a readability and comprehension point of view.

Sounds great! I might do that as well :)

goderbauer commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
+
+  @override
+  T resolve(Set<WidgetState> states) {
+    for (final MapEntry<WidgetStateMapKey, T> entry in map.entries) {
```
I am missing documentation stating in which order it will try to match the entries of the provided map. Can you add this to some prominent places?

nate-thegrate replied:
Good point. Since this is a private class's method, I'll document the behavior in each constructor that uses it.

goderbauer commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
+    } on TypeError {
+      throw ArgumentError(errorMessage);
+    }
+  }
```
I wonder if we shouldn't allow implicitly returning null. Would it maybe be better to explicitly require an entry for null in the map if you want it to ever resolve to null?

Would be interesting to look at some use cases to see how common the null case is and if making it explicit would just add a ton of boilerplate.

goderbauer replied:
Also, this implicit null behavior needs to be documented somewhere clearly.

nate-thegrate replied:
> this implicit null behavior needs to be documented somewhere clearly.

Absolutely!

goderbauer commented on packages/flutter/lib/src/widgets/widget_state.dart:
````diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
+/// Example:
+///
+/// ```dart
+/// WidgetStateProperty<Color?>.map(<WidgetStateMapKey, Color?>{
````
Alternatively: `WidgetStateProperty<Color?>.fromMap` to emphasize that this creates a  WidgetStateProperty from a map and doesn't map values to another value.

nate-thegrate committed:
rewrite documentation based on code review

nate-thegrate committed:
rename method & factories

nate-thegrate commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -8,6 +8,89 @@ import 'package:flutter/services.dart';
+/// one of its values, based on the first key that [isSatisfiedBy]
+/// the current set of states.
+///
+/// {@macro flutter.widgets.WidgetStateProperty.WidgetStateMap}
```
As far as the order of examples in this doc macro, my thoughts are that you don't _need_ to understand the operators in order to make a `WidgetStateMap`, but you definitely would need to understand the basics of a `WidgetStateMap` in order to know how to use the operators.

nate-thegrate commented on packages/flutter/lib/src/widgets/widget_state.dart:
````diff
@@ -8,6 +8,80 @@ import 'package:flutter/services.dart';
+  /// The key could also be created using one or more operators, for example:
+  ///
+  /// ```dart
+  /// final WidgetStateMapKey mapKey = WidgetState.focused | WidgetState.hovered;
````
After thinking it over for a while, I do really like the parallelism with `BoxConstraints`.

I'll go ahead and change the class name accordingly.

nate-thegrate committed:
change class name

goderbauer commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -8,6 +8,89 @@ import 'package:flutter/services.dart';
+///
+/// {@macro flutter.widgets.WidgetStateProperty.WidgetStateMap}
+abstract interface class WidgetStatesConstraint {
+  /// Whether the current [states] satisfy this object's criteria.
```
nit:
```suggestion
  /// Whether the provided [states] satisfy this object's criteria.
```

goderbauer commented on packages/flutter/lib/src/widgets/widget_state.dart:
```diff
@@ -519,12 +686,24 @@ class _WidgetStateTextStyle extends WidgetStateTextStyle {
+  /// the method throws an [ArgumentError].
+  ///
+  /// {@macro flutter.widgets.WidgetState.any}
+  const factory WidgetStateProperty.fromMap(WidgetStateMap<T> map) = _WidgetStateMapper<T>;
```
I suspect that this is a common entry point to WidgetState; maybe include the general example in this doc comment as well to give a general overview of how this all works.

goderbauer commented on packages/flutter/lib/src/widgets/widget_state.dart:
````diff
@@ -594,6 +780,95 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
+/// Example:
+///
+/// ```dart
+/// WidgetStateProperty<Color?>.fromMap(<WidgetStatesConstraint, Color?>{
````
Maybe leave a comment on here to explain that the map passed to fromMap is of type `WidgetStateMap<Color?>` for people wondering why this example is given on WidgetStateMap.

I agree with your comment above that assigning it to a variable is not a good idea.

goderbauer reviewed (approved):
LGTM after some minor doc adjustments (see comments).

nate-thegrate committed:
documentation tweaks

nate-thegrate commented:
Third time's the charm! 🤞
